[R] fix to-list and to-json functionality

mattpollock opened the pull request:
@Ramanth, @saigiridhar21, @wing328

This PR fixes an issue with converting R6 objects to JSON and R list types.

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package || exit
  ./bin/generate-samples.sh ./bin/configs/*.yaml || exit
  ./bin/utils/export_docs_generators.sh || exit
  ``` 
  (For Windows users, please run the script in [Git BASH](https://gitforwindows.org/))
  Commit all changed files. 
  This is important, as CI jobs will verify _all_ generator outputs of your HEAD commit as it would merge with master. 
  These must match the expectations made by your contribution. 
  You may regenerate an individual generator by passing the relevant config(s) as an argument to the script, for example `./bin/generate-samples.sh bin/configs/java*`. 
  IMPORTANT: Do **NOT** purge/delete any folders/files (e.g. tests) when regenerating the samples as manually written tests may be removed.
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master` (upcoming `7.x.0` minor release - breaking changes with fallbacks), `8.0.x` (breaking changes without fallbacks)
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

wing328 commented:
thanks for the PR

I ran the integration tests locally but got errors:
```
See chapter ‘Writing R documentation files’ in the ‘Writing R
Extensions’ manual.
* checking examples ... NONE
* checking for unstated dependencies in ‘tests’ ... OK
* checking tests ...
  Running ‘testthat.R’
 ERROR
Running the tests in ‘tests/testthat.R’ failed.
Last 13 lines of output:
   1. ├─testthat::expect_equal(nested_oneof$toJSONString(), "{\"size\":15,\"nested_pig\":{\"className\":\"BasquePig\",\"color\":\"red\"}}") at test_petstore.R:281:3
   2. │ └─testthat::quasi_label(enquo(object), label, arg = "object")
   3. │   └─rlang::eval_bare(expr, quo_get_env(quo))
   4. └─nested_oneof$toJSONString()
   5.   └─self$toList()
  ── Failure ('test_petstore.R:322:3'): Tests anyOf ──────────────────────────────
  danish_pig$toJSONString() not equal to original_danish_pig$toJSONString().
  Classes differ: 'character' is not 'json'
  ── Failure ('test_petstore.R:328:3'): Tests anyOf ──────────────────────────────
  basque_pig$toJSONString() not equal to original_basque_pig$toJSONString().
  Classes differ: 'character' is not 'json'

  [ FAIL 11 | WARN 5 | SKIP 107 | PASS 62 ]
  Error: Test failures
  Execution halted

* DONE
Status: 1 ERROR, 1 WARNING
See
  ‘/mnt/c/Users/wing3/Code/openapi-generator/samples/client/petstore/R/petstore.Rcheck/00check.log’
for details.
```

can you pleasea take a look?

to run the test, please run `mvn integration-test -f samples/client/petstore/R/pom.xml`

you will need to run the petstore server locally: https://github.com/OpenAPITools/openapi-generator/wiki/Integration-Tests

mattpollock commented:
Thanks @wing328. I had some trouble with the integration tests (my hosts file is locked down) but I think I fixed the issue. `jsonlite::toJSON` creates an object of type "json" instead of a character string, I converted back to a string to match the behavior in `master`.

mattpollock commented on modules/openapi-generator/src/main/resources/r/modelOneOf.mustache:
```diff
@@ -139,25 +139,35 @@
-    toJSONString = function() {
+    toJSONString = function(...) {
+      simple <- self$toSimpleType()
       if (!is.null(self$actual_instance)) {
-        as.character(jsonlite::minify(self$actual_instance$toJSONString()))
+        json <- jsonlite::toJSON(simple, auto_unbox = TRUE, ...)
+        return(as.character(jsonlite::minify(json)))
       } else {
-        NULL
+        return(NULL)
       }
     },
 
     #' @description
-    #' Serialize {{{classname}}} to JSON.
-    #'
-    #' @return JSON representation of the {{{classname}}}.
+    #' Convert to an R object. This method is deprecated. Use `toSimpleType()` instead.
     toJSON = function() {
+      .Deprecated(new = "toSimpleType", msg = "Use the '$toSimpleType()' method instead since that is more learly named. Use '$toJSONString()' to get a JSON string")
+      return(self$toSimpleType())
+    },
+
+    #' @description
+    #' Convert {{{classname}}} to a base R type
+    #'
+    #' @return A base R type, e.g. a list or numeric/character array.
+    toSimpleType = function() {
       if (!is.null(self$actual_instance)) {
-        self$actual_instance$toJSON()
+        return(self$actual_instance$toSimpleType())
       } else {
-        NULL
+        return(NULL)
       }
```
OK, I refactored this PR a bit since a realized that not all model types necessarily get transformed to R list types enroute to JSON.

The gist is that
1. I depreciated `toJSON` in favor of `toSimpleType` and made the behavior consistent (i.e., it returns an R list, or character vector, etc.). Previously `toJSON` might return a list or a JSON string, depending on the model type.
2. I added `toSimpleType` to all models
3. I modified `toJSONString` to use `toSimpleType()` to get the R object and then invoke the `jsonlite` transformations on that object (`toJSON` and `minifiy`).

wing328 commented:
thank you. let me test and see if i've any questions.

wing328 commented:
FYI. Filed follow-up PR https://github.com/OpenAPITools/openapi-generator/pull/20196 as it fails to run under WSL (Windows). (was running fine previously on Mac)

mattpollock commented on modules/openapi-generator/src/main/resources/r/modelGeneric.mustache:
```diff
@@ -203,10 +203,35 @@
-    #' @return {{{classname}}} in JSON format
+    #' Convert to an R object. This method is deprecated. Use `toSimpleType()` instead.
     toJSON = function() {
+      .Deprecated(new = "toSimpleType", msg = "Use the '$toSimpleType()' method instead since that is more learly named. Use '$toJSONString()' to get a JSON string")
```
Thanks - pushed a fix.

mattpollock commented:
I'm running/testing on a Mac too - thanks for handling the windows issue.

wing328 commented:
thanks for the bug fix

let's give it a try and see how that goes
